discussions/secure_web_api.md

```diff
@@ -0,0 +1,137 @@
+# L3AFD Secure web API
+
+## Overview
+
+Walmart developed L3AF to simplify the management and orchestration of multiple eBPF programs in an enterprise
+environment.
+
+L3AF’s control plane consists of multiple components that work together to orchestrate eBPF programs:
+
+- L3AF Daemon (L3AFD), which runs on each node where eBPF program runs. L3AFD reads configuration data and manages 
+  the execution and monitoring of eBPF programs running on the node.
+- Deployment APIs, which a user calls to generate configuration data. This configuration data includes which eBPF 
+  programs will run, their execution order, and the configuration arguments for each eBPF program.
+- A database and local KV store that stores the configuration data. 
+- A datastore that stores the eBPF program artifact (I.e., byte code and native code).
+
+When users want to deploy an eBPF program, they can call the L3AFD API with appropriate parameters. This request would
+generate a new config (KV pair). Once L3AFD reads this new config, it orchestrates eBPF programs on the Linux host
+as per the defined parameters. If the user gives a set of eBPF programs, then L3AFD can orchestrate all of them in the
+sequence that the user wanted (aka chaining).
+
+## Need for secure web APIs
+
+L3AFD implements a Go HTTPS client to download the configured eBPF programs from a datastore (package repository).
+However, this client does not support TLS. In the open-source world, users will presumably expect to be able
+to use TLS in this situation. This is also an early step toward using a secure eBPF Package Repository
+(https://github.com/l3af-project/l3afd/issues/2).
```

Comment:
I think the file-transfer mechanism should be decoupled from the local eBPF program lifecycle management mechanism, so that they could be in two separate daemons.  This would allow the file-transfer mechanism to alternatively be some existing mechanism like k8s or whatever else, while still allowing l3afd to manage the program chaining.  It would also allow removing code rather than adding TLS support per se, since one could use a daemon that already has TLS support and is more mature rather than the risk of bugs in new code.

Reply:
I think we should support both the l3afd download mechanism and other existing mechanisms.

Reply:
I think for now being able to specify local file path or https URL path to download eBPF programs meets the needs for now. Having a separate daemon for simple https download seems like overkill. If downloads and distribution becomes more complex we can reconsider a separate daemon.